File: agent/permissions/manager.py
from __future__ import annotations

import json
from typing import Any

from ..control.tools import make_pause_result

from .models import PermissionDecision
from .policy import PermissionPolicy


class PermissionManager:
    def __init__(self, control_manager):
        self.control_manager = control_manager
        self.policy = PermissionPolicy()
        self._approved_once: set[str] = set()
        self._denied_once: set[str] = set()

    def assess(self, tool_name: str, arguments: dict[str, Any] | None, *, registry=None) -> PermissionDecision:
        args = arguments or {}
        fingerprint = _fingerprint(tool_name, args)
        if fingerprint in self._approved_once:
            self._approved_once.remove(fingerprint)
            return PermissionDecision.allow(tool_name=tool_name, arguments=args)
        if fingerprint in self._denied_once:
            self._denied_once.remove(fingerprint)
            return PermissionDecision.deny(
                tool_name=tool_name,
                arguments=args,
                reason="user denied this high-risk operation",
            )
        return self.policy.assess(tool_name, args, self.control_manager.mode, registry=registry)

    def require_approval(
        self,
        decision: PermissionDecision,
        *,
        parent_call_id: str | None = None,
    ) -> str:
        interaction = self.control_manager.create_ask(
            questions=[
                {
                    "id": "permission",
                    "header": "权限",
                    "question": f"是否允许执行高风险操作 `{decision.tool_name}`？",
                    "options": [
                        {"label": "允许", "description": "批准本次操作，Agent 可继续执行。"},
                        {"label": "拒绝", "description": "不执行本次操作，让 Agent 改用更安全方案。"},
                    ],
                }
            ],
            context=self._context(decision),
            parent_call_id=parent_call_id,
            meta={
                "permission": {
                    "fingerprint": _fingerprint(decision.tool_name, decision.arguments or {}),
                    "tool_name": decision.tool_name,
                    "risk": decision.risk,
                    "reason": decision.reason,
                    "arguments": decision.arguments or {},
                }
            },
        )
        return make_pause_result(interaction.to_dict())

    def record_answer(self, interaction) -> None:
        permission = getattr(interaction, "meta", {}).get("permission") if getattr(interaction, "meta", None) else None
        if not isinstance(permission, dict):
            return
        fingerprint = str(permission.get("fingerprint") or "")
        if not fingerprint:
            return
        answer = interaction.answers.get("permission")
        choice = ""
        if isinstance(answer, dict):
            choice = str(answer.get("choice") or answer.get("freeform") or "")
        else:
            choice = str(answer or "")
        normalized = choice.strip().lower()
        if "允许" in normalized or "approve" in normalized or "allow" in normalized or "yes" == normalized:
            self._approved_once.add(fingerprint)
            self._denied_once.discard(fingerprint)
            return
        self._denied_once.add(fingerprint)
        self._approved_once.discard(fingerprint)

    @staticmethod
    def _context(decision: PermissionDecision) -> str:
        return "\n".join([
            "Permission Guard",
            f"risk: {decision.risk}",
            f"reason: {decision.reason}",
            f"tool: {decision.tool_name}",
            "arguments:",
            json.dumps(decision.arguments or {}, ensure_ascii=False, indent=2)[:1600],
        ])


def _fingerprint(tool_name: str, arguments: dict[str, Any]) -> str:
    try:
        encoded = json.dumps(arguments or {}, ensure_ascii=False, sort_keys=True, separators=(",", ":"))
    except (TypeError, ValueError):
        encoded = json.dumps(_json_safe(arguments or {}), ensure_ascii=False, sort_keys=True, separators=(",", ":"))
    return f"{tool_name}:{encoded}"


def _json_safe(value: Any) -> Any:
    if isinstance(value, dict):
        return {str(key): _json_safe(item) for key, item in value.items()}
    if isinstance(value, list):
        return [_json_safe(item) for item in value]
    if isinstance(value, (str, int, float, bool)) or value is None:
        return value
    return str(value)

File: packages/core/src/permissions/manager.ts
/**
 * PermissionManager (MIG-CTRL-017)。对齐 Python `agent/permissions/manager.py`。
 * approve/deny-once 指纹 + plan token；高风险 run_command 在 token 前先评估 (PE-13)。
 */
import type { ToolRegistry } from '../tools/registry'
import { makePauseResult } from '../control/tools'
import {
  PermissionDecision,
  stableJson,
  traceEntry,
  type PermissionTraceEntry,
  type PlanPermissionToken,
} from './models'
import { PermissionPolicy } from './policy'
import { isHighRiskCommand } from '../tools/resolvers'

/** PermissionManager 依赖的 ControlManager 表面。 */
export interface PermissionControlHost {
  readonly mode: string
  createAsk(opts: {
    questions: Array<Record<string, unknown>>
    context?: string
    parentCallId?: string | null
    meta?: Record<string, unknown> | null
  }): { toDict?: () => Record<string, unknown>; answers?: Record<string, unknown>; meta?: Record<string, unknown> } & Record<string, unknown>
  consumePlanPermissionToken?(opts: { toolName: string; arguments: Record<string, unknown> }): PlanPermissionToken | null
}

interface InteractionLike {
  meta?: Record<string, unknown>
  answers?: Record<string, unknown>
}

export class PermissionManager {
  private readonly controlManager: PermissionControlHost
  readonly policy: PermissionPolicy
  private readonly approvedOnce = new Set<string>()
  private readonly deniedOnce = new Set<string>()

  constructor(controlManager: PermissionControlHost) {
    this.controlManager = controlManager
    this.policy = new PermissionPolicy()
  }

  assess(toolName: string, args: Record<string, unknown> | null, opts?: { registry?: ToolRegistry | null }): PermissionDecision {
    const argv = args ?? {}
    const fingerprint = fingerprintOf(toolName, argv)
    if (this.approvedOnce.has(fingerprint)) {
      this.approvedOnce.delete(fingerprint)
      return PermissionDecision.allow({ toolName, arguments: argv, rule: 'user.approved_once' })
    }
    if (this.deniedOnce.has(fingerprint)) {
      this.deniedOnce.delete(fingerprint)
      return PermissionDecision.deny({
        toolName,
        arguments: argv,
        reason: 'user denied this high-risk operation',
        rule: 'user.denied_once',
      })
    }
    if (toolName === 'run_command' && isHighRiskCommand(String(argv.command ?? ''))) {
      return this.policy.assess(toolName, argv, this.controlManager.mode, { registry: opts?.registry ?? null })
    }
    const planDecision = this.planPermissionTokenDecision(toolName, argv)
    if (planDecision !== null) return planDecision
    return this.policy.assess(toolName, argv, this.controlManager.mode, { registry: opts?.registry ?? null })
  }

  requireApproval(decision: PermissionDecision, opts?: { parentCallId?: string | null }): string {
    const interaction = this.controlManager.createAsk({
      questions: [
        {
          id: 'permission',
          header: '权限',
          question: `是否允许执行高风险操作 \`${decision.toolName}\`？`,
          options: [
            { label: '允许', description: '批准本次操作，Agent 可继续执行。' },
            { label: '拒绝', description: '不执行本次操作，让 Agent 改用更安全方案。' },
          ],
        },
      ],
      context: this.context(decision),
      parentCallId: opts?.parentCallId ?? null,
      meta: {
        permission: {
          fingerprint: fingerprintOf(decision.toolName, decision.arguments ?? {}),
          tool_name: decision.toolName,
          risk: decision.risk,
          reason: decision.reason,
          rule: decision.rule,
          trace: decision.trace.map((item) => ({ rule: item.rule, outcome: item.outcome, detail: item.detail })),
          arguments: decision.arguments ?? {},
        },
      },
    })
    const dict = typeof interaction.toDict === 'function' ? interaction.toDict() : (interaction as unknown as Record<string, unknown>)
    return makePauseResult(dict)
  }

  recordAnswer(interaction: InteractionLike): void {
    const permission = interaction.meta && typeof interaction.meta === 'object' ? interaction.meta.permission : null
    if (!permission || typeof permission !== 'object') return
    const fingerprint = String((permission as Record<string, unknown>).fingerprint ?? '')
    if (!fingerprint) return
    const answer = interaction.answers?.permission
    let choice = ''
    if (answer && typeof answer === 'object') {
      choice = String((answer as Record<string, unknown>).choice ?? (answer as Record<string, unknown>).freeform ?? '')
    } else {
      choice = String(answer ?? '')
    }
    const normalized = choice.trim().toLowerCase()
    if (normalized.includes('允许') || normalized.includes('approve') || normalized.includes('allow') || normalized === 'yes') {
      this.approvedOnce.add(fingerprint)
      this.deniedOnce.delete(fingerprint)
      return
    }
    this.deniedOnce.add(fingerprint)
    this.approvedOnce.delete(fingerprint)
  }

  private context(decision: PermissionDecision): string {
    return [
      'Permission Guard',
      `risk: ${decision.risk}`,
      `rule: ${decision.rule}`,
      `reason: ${decision.reason}`,
      `tool: ${decision.toolName}`,
      'trace:',
      JSON.stringify(decision.trace.map((item) => ({ rule: item.rule, outcome: item.outcome, detail: item.detail })), null, 2).slice(0, 1200),
      'arguments:',
      JSON.stringify(decision.arguments ?? {}, null, 2).slice(0, 1600),
    ].join('\n')
  }

  private planPermissionTokenDecision(toolName: string, args: Record<string, unknown>): PermissionDecision | null {
    const consumer = this.controlManager.consumePlanPermissionToken
    if (typeof consumer !== 'function') return null
    const token = consumer.call(this.controlManager, { toolName, arguments: args })
    if (token === null || token === undefined) return null
    const trace: PermissionTraceEntry[] = [traceEntry('plan.permission_token', 'allow', `${token.planId}:${token.stepId}`)]
    return PermissionDecision.allow({ toolName, arguments: args, rule: 'plan.permission_token', trace })
  }
}

function fingerprintOf(toolName: string, args: Record<string, unknown>): string {
  return `${toolName}:${stableJson(args ?? {})}`
}
